more complete unit tests for state grains

dr4Ke opened the pull request:
This is a proposal for different, more complete unit tests for the state grains. I wrote it some days ago, hadn't notice that the current one existed. When I noticed, I wanted to add some tests I think are necessary, but the implementation is so different that I abandoned.

Please don't take this pull request in a bad way. It is not my intention to hurt anybody.

My implementation may be of less quality, as I'm more a sysadmin than a developer, but I think it has some needed additional tests, and I have a concern about the current implementation. I may be wrong, but I think it would be better to test the state result using the actual module functions, not mocked ones. Am I wrong? I always fear that a change introduced in the module functions could break the state behavior without being noticed.
I tried to follow the [Writing Unit Tests](http://docs.saltstack.com/en/latest/topics/development/tests/unit.html) page, using one function for each `return`.

There is one thing that bother me, in this tests. It's that I was forced to specify system accessible (but unix specific) paths for `__opts__['conf_file']` and `__opts__['cachedir']`. Is there a way to mock a file or directory, or a better way?

Please correct me if anything is wrong in my thinking.

Commit:
more complete unit tests for state grains

dr4Ke commented:
I have to add that these tests helps me working on a new implementation of `present` based on the new module function `grains.set`, which may solve issues #11870 and #15583.

jfindlay commented:
@dr4Ke, you make some excellent points.  Modules and states are typically closely related.  There seems to be [some failures](https://jenkins.saltstack.com/job/salt-pr-rs-ubuntu14.04-n/5775/) with the new tests.  Also this pull request cannot be merged without conflicts.  Also, I would like to have your new tests on 2015.5.  Would you mind rebasing on that branch?  Thanks.

dr4Ke commented:
@jfindlay The failures reported are on `__salt__['saltutil.sync_grains']()`. I don't have these errors when I run the tests myself. I fail to understand why it's happening.

jfindlay commented:
It means that `__salt__['saltutil.sync_grains']` needs to be mocked in the unit tests.

dr4Ke commented:
Yes. But I did. At least, I don't have trouble running these tests on my machine. And, there was no trouble in the unit tests I submitted for the `grains` module using the same technique. I used this decorator on the test class: `@patch.dict(grainsmod.__salt__, {'saltutil.sync_grains': MagicMock()})`. Is that not enough? Or should I use it for each function?

I can't reproduce the issue on my machines so I have trouble solving it.

Commit:
trying to fix jenkins error

There are many jenkins errors on linode CentOS7. The function `saltutil.sync_grains`
does not seem to be mocked correctly on this platform.

Based on https://github.com/jfindlay/salt/commit/1f5836d0bb4635de4bd0fe39fb8121b197e11cc1

dr4Ke commented:
I'm stuck there with the jenkins failures. They don't happen on my machine, and I don't know how to reproduce this issue.

Commit:
fix weird error in test test_absent_delete

In the function grains.setvals, when deleting a key (destructive=True),
the __grains__ var, containing the current grains was updated only if
the grains file actually contained the key. In this test, it was not the
case because the grains file is not actually used, but the current grains
var was set manually.

Commit:
unit tests for state grains: use a subdirectory for config file

This way, there is no /tmp/grains created on test execution. And the issue
with test_absent_delete is always triggered.
Maybe we should use mocked files, but I'm not sure how to do it.
